packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -0,0 +1,232 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* @flow */
+
+import React, { Component, type Element, type ComponentType } from 'react';
+import PropTypes from 'prop-types';
+import { cssModules } from 'bpk-react-utils';
+import omit from 'lodash/omit'
+import 'intersection-observer';
+
+import STYLES from './withInfiniteScroll.scss';
+
+const getClassNames = cssModules(STYLES);
+
+type scrollEvent = {
+  currentIndex: number,
+};
+
+type scrollFinishedEvent = {
+  totalNumberElements: number,
+};
+
+type onItemsFetchFunction = (
+  index: number,
+  elementsPerScroll: number,
+) => Promise<Array<any>>;
+
+export type Props = {
+  elementsPerScroll: number,
+  loadingComponent: ?Element<any>,
+  onItemsFetch: onItemsFetchFunction,
+  onClickSeeMore: (event: SyntheticEvent<HTMLButtonElement>) => void,
+  onScroll: (o: scrollEvent) => void,
+  onScrollFinished: (o: scrollFinishedEvent) => void,
+  seeMoreComponent: Element<any>,
+  seeMoreAfter: number,
+};
+
+export type State = {
+  index: number,
+  elementsToRender: Array<any>,
+  isListFinished: boolean,
+  showSeeMore: boolean,
+};
+
+type extendedProps = {
+  elements: Array<any>,
+};
+
+const withInfiniteScroll = (
+  ComponentToExtend: ComponentType<extendedProps>,
+): ComponentType<Props> => {
+  class Infinitize extends Component<Props, State> {
+    handleIntersection: IntersectionObserverCallback;
+    handleKeyPress: (e: SyntheticKeyboardEvent<HTMLButtonElement>) => void;
+    handleSeeMoreClick: (e: SyntheticEvent<HTMLButtonElement>) => void;
+    observer: IntersectionObserver;
+    onItemsFetch: onItemsFetchFunction;
+    onScroll: (o: scrollEvent) => void;
+    sentinel: ?HTMLElement;
+
+    static defaultProps;
+
+    constructor(props: Props) {
+      super(props);
+
+      this.state = {
+        index: 0,
+        elementsToRender: [],
+        isListFinished: false,
+        showSeeMore: false,
+      };
+
+      this.onItemsFetch = props.onItemsFetch;
+      this.onScroll = props.onScroll;
+      this.handleIntersection = async entries => {
```

Comment:
That was not what I meant about replacing it with class properties + arrow function. See this: https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-progress/src/BpkProgress.js#L136

This way you don't need the `bind` logic because it auto binds to the current instance. The same is valid for `handleSeeMoreClick` and `handleKeyPress`

Reply:
Ok, I didn't know that was legal 😄